### PR TITLE
Working Simulation in C++

### PR DESCRIPTION
**Pull request by NicholasFan235**
Added Simulation classes and bunch of other helpers.
Modifications across many classes.
Modified Place to allow linking people in different cells
Added reporter classes which monitor the simulation and output data to files.

### CODE REVIEW

File: cEpiabm/src/reporters/percell_compartment_reporter.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            InfectionStatus::Susceptible,
+            InfectionStatus::Exposed,
+            InfectionStatus::InfectASympt,
+            InfectionStatus::InfectMild,
+            InfectionStatus::InfectGP,
+            InfectionStatus::InfectHosp,
+            InfectionStatus::InfectICU,
+            InfectionStatus::InfectICURecov,
+            InfectionStatus::Dead,
+            InfectionStatus::Recovered
```
Comment (KCGallagher) [inline cEpiabm/src/reporters/percell_compartment_reporter.cpp:23]: I don't like that we have to define these in the `cell_compartment_reporter` and the `percell_compartment_reporter` - is there any way to inherit/import these from a single place (so that we would only have to change them in one place if we wanted to change the list of compartments

File: cEpiabm/src/simulations/basic_simulation.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    };
+
+    typedef std::shared_ptr<BasicSimulation> BasicSimulationPtr;
+
+}
+
+
+
+
+#endif // EPIABM_SIMULATINS_BASIC_SIMULATION_HPP
```
Comment (KCGallagher) [inline cEpiabm/src/simulations/basic_simulation.hpp:47]:
```suggestion


#endif // EPIABM_SIMULATINS_BASIC_SIMULATION_HPP

```

Ditto on empty lines

File: cEpiabm/src/sweeps/household_sweep.cpp [hunk 3]
```diff
@@ -78,7 +78,7 @@ namespace epiabm
 
         double infectiousness = Covidsim::CalcHouseInf(infector, timestep);
         double susceptibility = Covidsim::CalcHouseSusc(infector, infectee, timestep);
-        double foi = infectiousness * susceptibility;
+        double foi = infectiousness * susceptibility * 0.5;
 
         if (static_cast<double>(std::rand() % 1000000) / static_cast<double>(1000000) < foi)
```
Comment (KCGallagher) [inline cEpiabm/src/sweeps/household_sweep.cpp:83]: I am confused about the factor of 1/2 here - this isn't currently consistent with pyEpiabm but should we add it there?

And why the use of a million in L83?

File: cEpiabm/test/dataclasses/test_person.cpp [hunk 4]
```diff
@@ -40,3 +46,29 @@ TEST_CASE("dataclasses/person: test setHousehold", "[Person]")
         static_cast<size_t>((std::rand()%500) * 2 + 1)) == false);
```
Comment (KCGallagher) [inline cEpiabm/test/dataclasses/test_person.cpp:46]: Does that rely on it producing the same random number as the previous call?

File: cEpiabm/test/dataclasses/test_place.cpp [hunk 5]
```diff
@@ -1,25 +1,149 @@
...
+        subject.cells().push_back(Cell(i));
+        subject.cells()[i].people().reserve(n_people);
+        for (size_t j = 0; j < n_people; j++)
+        {
+            subject.cells()[i].people().push_back(Person(i, j, j));
+
+            /*printf("Microcell: %ld, Person: (%ld, %ld)\n",
+                i,
+                subject.people()[subject.microcells()[i].people()[j]].cellPos(),
+                subject.people()[subject.microcells()[i].people()[j]].microcellPos());//*/
```
Comment (KCGallagher) [inline cEpiabm/test/dataclasses/test_place.cpp:32]:
```suggestion
```

Fine for testing but I think commented out code is generally not a good thing?

**Review by KCGallagher — COMMENTED** [review]
Looks really good - I have added minor comments but mostly on things I don't understand - apologies for the stupid questions!

P.S There is an empty file `simulation_interface.hpp` - I can't comment on it directly because it has no content but maybe this should be deleted?
P.P.S When you comment `namespace epiabm` at the final closing bracket, is that just good practice to document the extent of code using this namespace?

File: cEpiabm/examples/basic_example/main.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+#include "sweeps/new_infection_sweep.hpp"
+#include "sweeps/basic_host_progression_sweep.hpp"
+#include "sweeps/random_seed_sweep.hpp"
+#include "reporters/cell_compartment_reporter.hpp"
+#include "reporters/percell_compartment_reporter.hpp"
+#include "reporters/population_compartment_reporter.hpp"
+
+#include <random>
+
+void run()
```
[the remark this answers is not in the document]
Comment (NicholasFan235) [reply cEpiabm/examples/basic_example/main.cpp:17]: More of a testing script, in the future probably going to expose it to python. Could comment here, or in a jupyter notebook might be better

File: cEpiabm/examples/basic_example/main.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+{
+    using namespace epiabm;
+
+    std::srand(100); // random seed
```
[the remark this answers is not in the document]
Comment (NicholasFan235) [reply cEpiabm/examples/basic_example/main.cpp:21]: Right now it sets it for everything, but this probably should be changed in the future. Currently all the random number generation is just from uniform distributions, need to create some random number classes probably.

File: cEpiabm/examples/basic_example/main.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    LogFile::Instance()->configure(2, std::filesystem::path("output/log.log"));
+
+    // Make Population and Link Households
+    PopulationPtr population = PopulationFactory().makePopulation(10, 10, 1000);
+    HouseholdLinker().linkHouseholds(population, 1, 100);
+    population->initialize();
+
+    // Randomly Seed Population with Infections
+    {
+        RandomSeedSweep randomizer = RandomSeedSweep(600);
```
[the remark this answers is not in the document]
Comment (NicholasFan235) [reply cEpiabm/examples/basic_example/main.cpp:33]: I'm not sure named parameters are in the C++ standard. I think some compilers are okay with it depending on parameters. Seems like GCC supports it from C++20, but MSVC you need to add extra flags. You'd normally add those as a comment

File: cEpiabm/src/reporters/percell_compartment_reporter.cpp (hunk 1 above)
Comment (NicholasFan235) [reply cEpiabm/src/reporters/percell_compartment_reporter.cpp:23]: Could create a global variable which stores a list of all statuses.

File: cEpiabm/src/sweeps/basic_host_progression_sweep.cpp [hunk 9]
```diff
@@ -95,7 +95,7 @@ namespace epiabm
      */
     inline unsigned short BasicHostProgressionSweep::time_to_next_status(Person* /*person*/, InfectionStatus /*status*/)
     {
-        return static_cast<unsigned short>(std::rand() % 10 + 1);
+        return static_cast<unsigned short>(std::rand() % 14 + 14);
```
[the remark this answers is not in the document]
Comment (NicholasFan235) [reply cEpiabm/src/sweeps/basic_host_progression_sweep.cpp:98]: Just an arbitrary random number for now. I don't think it will be a uniform distribution in the future, so parameters don't really matter

File: cEpiabm/src/sweeps/household_sweep.cpp (hunk 3 above)
Comment (NicholasFan235) [reply cEpiabm/src/sweeps/household_sweep.cpp:83]: I think covid-sim had that factor, could be wrong. Used million to generate random float to 6dp, in the future I imagine this will be handled by some random number classes elsewhere

File: cEpiabm/test/dataclasses/test_person.cpp (hunk 4 above)
Comment (NicholasFan235) [reply cEpiabm/test/dataclasses/test_person.cpp:46]: No, checking setHousehold returns false if the person is already linked to a household. (easier not to allow changing household)

File: cEpiabm/test/dataclasses/test_place.cpp (hunk 5 above)
Comment (NicholasFan235) [reply cEpiabm/test/dataclasses/test_place.cpp:32]: Fine to have in the tests, this bit of code is pretty tricky, and its possible we'd have to modify it in the future so it'll be useful to have these prints handy. Very easy to mess up initialization and get segmentation faults with this function.

File: cEpiabm/src/reporters/percell_compartment_reporter.cpp (hunk 1 above)
Comment (KCGallagher) [reply cEpiabm/src/reporters/percell_compartment_reporter.cpp:23]: Yeah that definitely seems tidier and less error-prone/easier for the future if that's easy to do!

File: cEpiabm/src/sweeps/household_sweep.cpp (hunk 3 above)
Comment (KCGallagher) [reply cEpiabm/src/sweeps/household_sweep.cpp:83]: Ooh okay - will have a look as the python doesn't have this so one backend needs to be changed!

File: cEpiabm/test/dataclasses/test_place.cpp (hunk 5 above)
Comment (KCGallagher) [reply cEpiabm/test/dataclasses/test_place.cpp:32]: Sorry I didn't notice this was a test file - all good!

File: cEpiabm/src/sweeps/household_sweep.cpp (hunk 3 above)
Comment (KCGallagher) [reply cEpiabm/src/sweeps/household_sweep.cpp:86]: @NicholasFan235 have been going through the CovidSim code on this and I really can't find this factor of 1/2 - have you got a reference for it?

File: cEpiabm/src/covidsim.cpp [hunk 10]
```diff
@@ -25,6 +41,14 @@ namespace epiabm
...
+    double Covidsim::CalcSpaceSusc(
+        Cell* /*cell*/,
+        Person* /*infectee*/,
+        unsigned short int )
+    {
+        return 0.2;
+    }
 
 
 } // namespace epiabm
```
Comment (KCGallagher) [inline cEpiabm/src/covidsim.cpp:54]:
```suggestion
} // namespace epiabm

```
Seems to want newline characters at the end of files

File: cEpiabm/src/dataclasses/infection_status.hpp [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+            case InfectionStatus::InfectICU: return "InfectICU";
+            case InfectionStatus::InfectICURecov: return "InfectICURecov";
+            case InfectionStatus::Recovered: return "Recovered";
+            case InfectionStatus::Dead: return "Dead";
+            default: return "Unknown";
+        }
+    }
+}
+
+#endif // EPIABM_DATACLASSES_INFECTION_STATUS_HPP
```
Comment (KCGallagher) [inline cEpiabm/src/dataclasses/infection_status.hpp:42]:
```suggestion
#endif // EPIABM_DATACLASSES_INFECTION_STATUS_HPP

```
Another warning for missing end-line character - not wanting to self publicise but I think setting up the `.editorconfig` file could avoid these?

File: cEpiabm/src/dataclasses/microcell.cpp [hunk 12]
```diff
@@ -1,27 +1,30 @@
...
         m_households(),
-        m_cellPos(cellPos)
+        m_cellPos(cellPos),
+        m_compartmentCounter()
     {}
 
+    Microcell::~Microcell()
+    {
+        //std::cout << "Microcell Destructor" << std::endl;
+    }
```
Comment (KCGallagher) [inline cEpiabm/src/dataclasses/microcell.cpp:18]: Do we still need this (I know you have it in a few files, like `person.cpp` as well)?

File: cEpiabm/src/dataclasses/person.hpp [hunk 13]
```diff
@@ -41,18 +34,20 @@ namespace epiabm
...
+        std::set<size_t> m_places; // Indices of Places which the person is a member of;
 
     public:
         Person(size_t microcell, size_t cellPos, size_t mcellPos);
-        ~Person() = default;
-        Person(const Person&) = default;
-        Person(Person&&) = default;
+        ~Person();
+        //Person(const Person&) = default;
+        //Person(Person&&) = default;
```
Comment (KCGallagher) [inline cEpiabm/src/dataclasses/person.hpp:45]: Should these still be commented?

File: cEpiabm/src/population_factory.cpp [hunk 14]
```diff
@@ -36,7 +36,7 @@ namespace epiabm
 
     void PopulationFactory::addCell(PopulationPtr population)
     {
-        population->cells().push_back(Cell());
+        population->cells().emplace_back(population->cells().size());
```
Comment (KCGallagher) [inline cEpiabm/src/population_factory.cpp:39]: Not really a comment but more a learning thing for me - why is this better than `push_back()`?

File: cEpiabm/src/python_bindings/bind_dataclasses.hpp [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        .def_readwrite("susceptibility", &PersonParams::susceptibility)
+        .def_readwrite("infectiousness", &PersonParams::infectiousness)
+        .def_readonly("next_status_time", &PersonParams::next_status_time);
+    py::class_<Person, PersonPtr>(m, "Person")
+        .def("status", &Person::status, py::return_value_policy::copy)
+        .def("params", &Person::params, py::return_value_policy::reference)
+        .def("cell_pos", &Person::cellPos, py::return_value_policy::copy)
+        .def("microcell_pos", &Person::microcellPos, py::return_value_policy::copy)
+        .def("microcell", &Person::microcell, py::return_value_policy::copy)
+        .def("places", &Person::places, py::return_value_policy::reference)
```
Comment (KCGallagher) [inline cEpiabm/src/python_bindings/bind_dataclasses.hpp:41]: Again a learning thing, but what is the distinction between the different `return_value_policy` values here?

File: cEpiabm/src/simulations/basic_simulation.cpp [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+                    reporter->report(m_population, timestep);
+            }
+        //}
+        // LCOV_EXCL_START
+        /*
+        catch (std::exception& e)
+        {
+            LOG << LOG_LEVEL_ERROR << "Error iterating through timesteps: " << e.what();
+        }*/
+        // LCOV_EXCL_END
```
Comment (KCGallagher) [inline cEpiabm/src/simulations/basic_simulation.cpp:56]: Was this for testing or does it have some function?

File: cEpiabm/src/sweeps/spatial_sweep.cpp [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+namespace epiabm
+{
+
+    SpatialSweep::SpatialSweep() {}
+
+    void SpatialSweep::operator()(const unsigned short timestep)
+    {
+        LOG << LOG_LEVEL_DEBUG << "Beginning Spatial Sweep " << timestep;
+        if (m_population->cells().size() <= 1){
+            return;  // no intercell infections if only one cell
```
Comment (KCGallagher) [inline cEpiabm/src/sweeps/spatial_sweep.cpp:21]: I wonder if clauses like this should be logged - I think that could be quite useful!

**Review by KCGallagher — COMMENTED** [review]
Remember us saying the other day we wanted to get this merged in, so thought I would rereview to cover all the new parts I didn't see last time! All looks good, have added a couple of comments to clarify things (and also to help me understand the code as well!)

File: cEpiabm/src/sweeps/household_sweep.cpp (hunk 3 above)
Comment (NicholasFan235) [reply cEpiabm/src/sweeps/household_sweep.cpp]: You're right, 0.5 factor just appears in a later commit, probably from testing

File: cEpiabm/src/dataclasses/person.hpp (hunk 13 above)
Comment (NicholasFan235) [reply cEpiabm/src/dataclasses/person.hpp:45]: Commented because they're unused and are technically source implementations in the header file. Not sure what the workaround is, but these lines were causing whole repository to be recompiled even if unrelated files were changed.

File: cEpiabm/examples/basic_example/main.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (NicholasFan235) [reply cEpiabm/examples/basic_example/main.cpp]: Okay, I see what you mean. It's not a keyword argument, just a comment.
Something like this is done in some of the classes where there is currently an unused parameter and the compiler doesn't like that. So the future parameter is just commented out.

File: cEpiabm/src/python_bindings/bind_dataclasses.hpp (hunk 15 above)
Comment (NicholasFan235) [reply cEpiabm/src/python_bindings/bind_dataclasses.hpp:41]: Normally return_value_policy is about ownership i.e. whether python or c++ will delete the object when there are no references to it.
Copy creates a new object in Python with the same value as in C++, and Python has full ownership.
Reference means Python can look at and modify the C++ variable, but when Python is finished with it it won't try to delete it (since it's still a C++ owned object)

There is also return_value_policy::take_ownership
Here the C++ object is given to Python, but Python takes responsibility for deleting it when it's finished with it.

File: cEpiabm/src/simulations/basic_simulation.cpp (hunk 16 above)
Comment (NicholasFan235) [reply cEpiabm/src/simulations/basic_simulation.cpp:56]: Undecided. Is better to catch the error so it can be logged. But for testing with a debugger I needed the error to reach the top level and cause a crash.
TLDR: Was for testing

**Review by I-Bouros — APPROVED** [review]
I think it is okay, probably more documentation/comments would be nicer in places though.

File: cEpiabm/src/population_factory.cpp (hunk 14 above)
Comment (NicholasFan235) [reply cEpiabm/src/population_factory.cpp:39]: push_back would construct the Cell (as an rvalue), then copy it into the vector.
emplace_back constructs the Cell in place.

File: cEpiabm/src/simulations/basic_simulation.cpp (hunk 16 above)
Comment (KCGallagher) [reply cEpiabm/src/simulations/basic_simulation.cpp:56]: Does that mean we don't need it anymore, or is it still used in the automated tests? Just aware that's quite a public facing file so commented out code like that might be confusing!

File: cEpiabm/src/household_linker.hpp [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+    class HouseholdLinker
+    {
+    private:
+    public:
+
+        /**
+         * @brief Link households
+         * Create n_households per microcell.
+         * Person has percInHousehold to be in a household.
```
Comment (rcw5890) [inline cEpiabm/src/household_linker.hpp:20]:
```suggestion
         * Person has probability percInHousehold (expressed as percentage) to be in a household.
```

File: cEpiabm/src/simulations/basic_simulation.hpp [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+        void setup();
+        void teardown();
+    };
+
+    typedef std::shared_ptr<BasicSimulation> BasicSimulationPtr;
+
+}
+
+
+#endif // EPIABM_SIMULATINS_BASIC_SIMULATION_HPP
```
Comment (rcw5890) [inline cEpiabm/src/simulations/basic_simulation.hpp:45]:
```suggestion
#endif // EPIABM_SIMULATIONS_BASIC_SIMULATION_HPP
```

**Review by rcw5890 — CHANGES_REQUESTED** [review]

**Review by rcw5890 — APPROVED** [review]

**Review by KCGallagher — APPROVED** [review]

**Comment by KCGallagher** [comment]
@NicholasFan235 Is this ready to be merged then?
